Add compiler-generated code tests for NativeAot

sbomer opened the pull request:
This makes progress towards https://github.com/dotnet/runtime/issues/68786. It is still hitting issues such as https://github.com/dotnet/runtime/issues/68688 and https://github.com/dotnet/runtime/issues/77455, which are referenced in the test comments.

Also fixes an issue where warnings were not produced for ldftn. The test changes are a modified copy of tests from the linker codebase. See https://github.com/dotnet/linker/pull/3085 for the included changes.

sbomer committed:
Add compiler-generated code tests for NativeAot

Also fixes an issue where warnings were not produced for ldftn.

tlakollo reviewed (approved):
Looks good to me. Nice catch on the Release vs Debug releases. Did you try the checked version too? I'm guessing it behaves like the Debug one

sbomer committed:
Add Release mode test

sbomer committed:
Remove debug output from ResultChecker

sbomer committed:
Fix test for analyzer

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs:
```diff
@@ -482,8 +482,8 @@ protected virtual void Scan(MethodIL methodBody, ref InterproceduralState interp
                             if (methodBody.GetObject(reader.ReadILToken()) is MethodDesc methodOperand)
                             {
                                 TrackNestedFunctionReference(methodOperand, ref interproceduralState);
+                                HandleCall(methodBody, methodOperand, opcode, offset, new ValueNodeList() { MultiValueLattice.Top }, out MultiValue _);
```
This doesn't match MethodBodyScanner.cs under ReferenceSource (or in the linker repo main):

https://github.com/dotnet/linker/blob/4db6ac94ffa8bf6dae8ad9c5c68f94b58de917fd/src/linker/Linker.Dataflow/MethodBodyScanner.cs#L391-L394

Do we understand why linker doesn't need it and are we comfortable with a diff? Cc @vitek-karas

MichalStrehovsky commented:
> I don't think there is a checked configuration for managed code by default

We do support building the tools directory as checked:

https://github.com/dotnet/runtime/blob/7b5ab359a84d6a2c36a864a23c23bbe7686cae95/src/coreclr/tools/Directory.Build.props#L11-L15

It may become less relevant if linker tests get moved out of the coreclr test tree in the future (we don't need to support checked outside the coreclr tree).

sbomer committed:
Merge remote-tracking branch 'origin/main' into cgcFixes

sbomer committed:
Revert ldftn change

vitek-karas commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/Mono.Linker.Tests.Cases.csproj:
```diff
@@ -8,6 +8,8 @@
     <WarningLevel>0</WarningLevel>
     <AnalysisLevel>0</AnalysisLevel>
+    <RunAnalyzers>true</RunAnalyzers>
+    <DefineConstants>$(DefineConstants);NATIVEAOT</DefineConstants>
```
I know this is handy when writing the code, but do we need this to build the tests?
My point is that once the source trees are shared, there should be no reason to build the test cases assembly twice, since it actually isn't used as-is. We rebuild each test separately at runtime - and that build already defines `NATIVEAOT`.

vitek-karas commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/RequiresCapability/RequiresInCompilerGeneratedCode.cs:
```diff
@@ -0,0 +1,2287 @@
+{
+	[SkipKeptItemsValidation]
+	[ExpectedNoWarnings]
+#if NATIVEAOT
+	[Define ("NATIVEAOT")]
+#endif
```
The infra should already do this:
https://github.com/dotnet/runtime/blob/8606a915f57d6e1d68b7e2c0fe788bc4180eed01/src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/TestRunner.cs#L77-L78

vitek-karas commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/RequiresCapability/RequiresInCompilerGeneratedCode.cs:
```diff
@@ -0,0 +1,2287 @@
+			// In release mode, the async state machine is a struct which doesn't have a constructor, so no warning is emitted.
+#endif
+			// Linker warns about reflection access to compiler-generated state machine members.
+			// TODO: implement this for NativeAot (https://github.com/dotnet/runtime/issues/68786)
```
If the work is tracked by the issue I would remove the "TODO", just leave the link to the bug

vitek-karas commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/RequiresCapability/RequiresInCompilerGeneratedCode.cs:
```diff
@@ -0,0 +1,2287 @@
+			[ExpectedWarning ("IL2026", "--TestLocalFunctionInMethodWithRequires--", ProducedBy = ProducedBy.Trimmer)]
+			[ExpectedWarning ("IL2026", "--TestLocalFunctionWithClosureInMethodWithRequires--", ProducedBy = ProducedBy.Trimmer)]
+			// Linker warns about reflection access to compiler-generated code
+			// TODO: implement this for NativeAot (https://github.com/dotnet/runtime/issues/68786)
```
Same as above - just remove the TODO since it's already tracked.

vitek-karas commented on src/coreclr/tools/aot/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs:
```diff
@@ -220,8 +214,9 @@ private void VerifyLoggedMessages (AssemblyDefinition original, TestLogWriter lo
 										}
-										if (actualName.StartsWith (expectedMember.DeclaringType.FullName) &&
-											actualName.Contains (".cctor") && (expectedMember is FieldDefinition || expectedMember is PropertyDefinition)) {
+										if (actualName.StartsWith ($"[test]{expectedMember.DeclaringType.FullName.Replace ("/", ".")}") &&
+											actualName.Contains (".cctor") &&
+											(expectedMember is FieldDefinition || expectedMember is PropertyDefinition)) {
```
Ideally we should rewrite this to use GetExpectedOriginDisplayName and GetActualOriginDisplayName - look at LogMessageHasSameOriginMember where we compare Cecil and Internal.TypeSystem members. We basically rely on the display name which is currently "C# like" and on the code on both sides to adapt the specific type systems to it.

It tends to work pretty well.

vitek-karas reviewed (approved):
Couple of comments, but nothing blocking.

sbomer committed:
PR feedback, fix display name helpers

- Remove TODO comments
- Use built-in NATIVEAOT define
- Use display name helpers
